### ISSUE TITLE
Extending SARIMAXResults ignores DatetimeIndex

### ISSUE THREAD

**Issue by alexsmartens** [statsmodels/statsmodels#6301]
#### Describe the bug
Extending SARIMAXResults ignores DatetimeIndex and throws the following warning:
```
/home/user/Programs/anaconda3/envs/vision/lib/python3.7/site-packages/statsmodels/tsa/base/tsa_model.py:219: ValueWarning: A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.
  ' ignored when e.g. forecasting.', ValueWarning)
```

#### Code Sample, a copy-pastable example if possible
data: [my_data_sample2.csv](https://drive.google.com/file/d/185CkaiFpeKsy978PnawqoclwVS3tIOzQ/view?usp=sharing)

```python
endog = pd.read_csv("my_data_sample2.csv", index_col=0, parse_dates=True)

n_sample = 500
endog_train = endog.iloc[:n_sample]
endog_test = endog.iloc[n_sample:n_sample*2]

mod = sm.tsa.SARIMAX(endog, order=(1, 0, 0), trend='c')
res = mod.fit(disp=False)

endog_new_point = endog_test[0:1 + 1]
res = res.extend(endog_new_point)
# res = res.append(endog_new_point, refit=False)
pred_res_all = res.get_forecast(steps=2)
pred_means = pred_res_all.predicted_mean

print(pred_means)
```
Output:

```
2    100.838670
3    100.838663
dtype: float64
```
If the issue has not been resolved, please file it in the issue tracker.

#### Expected Output
The output index is expected to be datetime, the same as indices produced by `append`.
```
2019-04-24 22:50:40    100.838670
2019-04-24 22:51:40    100.838663
Freq: T, dtype: float64
```

Code that generated this output:

```python
endog = pd.read_csv("my_data_sample2.csv", index_col=0, parse_dates=True)

n_sample = 500
endog_train = endog.iloc[:n_sample]
endog_test = endog.iloc[n_sample:n_sample*2]

mod = sm.tsa.SARIMAX(endog, order=(1, 0, 0), trend='c')
res = mod.fit(disp=False)

endog_new_point = endog_test[0:1 + 1]
# res = res.extend(endog_new_point)
res = res.append(endog_new_point, refit=False)
pred_res_all = res.get_forecast(steps=2)
pred_means = pred_res_all.predicted_mean

print(pred_means)
```

#### Output of ``import statsmodels.api as sm; sm.show_versions()``

`0.11.0.dev0+628.g17f7c3481`

**Comment by ChadFulton** [comment]
Basically the issue here is that the date index doesn't have a frequency. The model infers the frequency to be minutes, but then the new dataset passed to `extend` is too short for Pandas to infer the frequency, so it just considers it to be an unsupported index.

I think the solution here will be to raise an exception if the new data passed to `extend` (or `append`) doesn't have an index that can be matched to the end of the model's index. At least then users will have clarity about the solution to the problem.

**Comment by alexsmartens** [comment]
Thanks for the explanation. I added frequency and it works now.
The problem is pandas does not infer frequency by default, this apparently causes some confusion.

> Basically the issue here is that the date index doesn't have a frequency. The model infers the frequency to be minutes, but then the new dataset passed to extend is too short for Pandas to infer the frequency, so it just considers it to be an unsupported index.

In my understanding both training and testing sets **should** absolutely have the same frequency, correct? 

If this is the case, then it seems like it might be logical to use the **frequency from the training set** when the new dataset passed to `extend` is too short for Pandas to infer the frequency, what do you think?

**Comment by bashtage** [comment]
That starts feeling like a bit magic.  PEP-020: Explicit is better than implicit.

**Comment by ChadFulton** [comment]
> In my understanding both training and testing sets should absolutely have the same frequency, correct?

> That starts feeling like a bit magic. PEP-020: Explicit is better than implicit.

With the combination of these two comments, I think that we are on the right path with #6334. We now raise an error if the given frequency from the test set disagrees with that from the training set.

> If this is the case, then it seems like it might be logical to use the frequency from the training set when the new dataset passed to extend is too short for Pandas to infer the frequency, what do you think?

In addition to wanting to avoid too much magic, we also don't have the resources to handle / maintain a bunch of special cases. Practically speaking, I don't think we have any other option but to delegate these kinds of things to Pandas, and this means we have to require that users provide data with indexes that are straightforward to extend.

**Comment by alexsmartens** [comment]
I also noticed that I get false frequency warnings when using pandas `asfreq` and `infer_freq`, example:
```

endog = pd.read_csv("my_data_sample2.csv", index_col=0, parse_dates=True)

n_sample = 500
endog_train = endog.iloc[:n_sample]
endog_train = endog_train.asfreq(pd.infer_freq(endog_train.index))
endog_test = endog.iloc[n_sample:n_sample*2]
endog_test = endog_test.asfreq(pd.infer_freq(endog_test.index))

mod = sm.tsa.SARIMAX(endog, order=(1, 0, 0), trend='c')
res = mod.fit(disp=False)

print(endog_test[0:1 + 1].index)
endog_new_point = endog_test[0:1 + 1]
res = res.extend(endog_new_point)
# res = res.append(endog_new_point, refit=False)
pred_res_all = res.get_forecast(steps=2)
pred_means = pred_res_all.predicted_mean

print(pred_means)
```

**Comment by ChadFulton** [comment]
I'm guessing that may be because you're passing `endog` to SARIMAX, but maybe you meant to pass `endog_train`?

**Comment by alexsmartens** [comment]
You're right, thanks!
